### PR TITLE
Added 'To, Cc and Bcc' fields for mailing an invoice

### PR DESCRIPTION
**Pull request by mfrausto-mdw**
Overview
----------------------------------------
Added functionality and altered templates to add 'To, Cc and Bcc' fields for mailing an invoice

Before
----------------------------------------
mailing_invoice_before

After
----------------------------------------
mailing_invoice_after

Technical Details
----------------------------------------
-Added toEmails property to `CRM/Contact/Form/Task/EmailCommon.php`
-Modified ` CRM/Contribute/Form/Task/Invoice.php` file in order to add the new form elements and rules for them
-Added helper function into `CRM/Core/BAO/Email.php`
-Added help assets to `templates/CRM/Contact/Form/Task/Email.hlp`
-Modified template ` templates/CRM/Contribute/Form/Task/Invoice.tpl` in order to show fields added

Comments
----------------------------------------
https://mydropwizard.atlassian.net/browse/ROUND-669

### CODE REVIEW

**Comment by mfrausto-mdw** [comment]
I've created this one as a continuity of this one:

https://github.com/civicrm/civicrm-core/pull/12889

@eileenmcnaughton

File: CRM/Core/BAO/Email.php
Comment (dsnopek) [inline CRM/Core/BAO/Email.php]: This bit is wrong - it's grabbing the contact for the currently logged in user, NOT the user we're trying to email to. This method should probably take an argument with the contact ID for who we want to email.

**Comment by dsnopek** [comment]
I noticed another fundamental issues in this which I didn't notice previously - sorry!

`CRM_Contribute_Form_Task_Invoice` can be used to send multiple emails as well. When doing multiple emails at once, it should show just the location type (so, you could send send multiple invoices to the "Billing" email) where it falls back on the primary email if that contact doesn't have an email of that type.

I that accommodating that will require making a lot of changes to the current PR...

**Comment by dsnopek** [comment]
@mfrausto-mdw I just pushed a commit which fixed the first issue, but not the second.

**Comment by mfrausto-mdw** [comment]
@dsnopek I pushed some changes regarding the second issue.

File: CRM/Core/BAO/Email.php
Comment (dsnopek) [inline CRM/Core/BAO/Email.php]: Hrm. I think this is not quite right. This appears to be setting up the email to go to both emails (ie. 'email1@example.com, email2@example.com'), rather than 2 individual emails to each of the recipients.

Comment (dsnopek) [reply CRM/Core/BAO/Email.php]: This function (if it's even necessary) should just return the list of location, and not concrete email addresses

File: CRM/Contribute/Form/Task/Invoice.php
Comment (dsnopek) [reply CRM/Contribute/Form/Task/Invoice.php]: See comment below. Here we should be getting just the location type from the form, and then use that to get the real email address from `$contribution->contact_id`

**Review by dsnopek — CHANGES_REQUESTED** [review]
Unfortunately, I think this still isn't quite right...

If I want to send contribution emails to multiple contacts, what I expect to happen is that multiple _individual_ emails will be sent to each contact. This is important because each individual email will have a different invoice attached to it (and we don't want contacts to get invoices that are meant for other contacts). This code currently appears to now send all the emails to all the selected contacts.

So, overall, I think what this means is that the form shouldn't select a concrete set of emails, just the location type. And then the location type will be converted to the correct email for that particular email when building the email template.

**Comment by mfrausto-mdw** [comment]
@dsnopek Comments addressed, please let me know if all looks good.

**Review by dsnopek — APPROVED** [review]
Read the code, and it looks functionally correct! I did some minimal manual testing, and it seems to be behaving correctly too, although, I didn't actually send any emails, just confirmed the values on the forms looked right. I think it's time to remove the "WIP" from the title and see if we can get some review from the maintainers :-)

**Comment by mfrausto-mdw** [comment]
Ok, removed WIP and ready for review.

**Comment by eileenmcnaughton** [comment]
@mattwire @seamuslee001 any chance one of you could review this? @mfrausto-mdw have you written any tests for Civi before?

**Comment by eileenmcnaughton** [comment]
@mfrausto-mdw can you please rebase this

**Comment by eileenmcnaughton** [comment]
Looks like this has been needing a rebase for a while - I'm going to close it & let you re-open when it is reviewable

**Pull request closed without merging**
